R/init.R

```diff
@@ -153,6 +153,16 @@ init <- function(data,
   # log
   teal.logger::log_system_info()
 
+  if (!identical(Sys.getenv("TESTTHAT"), "true")) {
+    future::plan(future::multisession, workers = 2)
```

Comment:
is this really necessary?

Moreover, note this:

> Please refrain from modifying the future strategy inside your packages / functions, i.e. do not call plan() in your code. Instead, leave the control on what backend to use to the end user. This idea is part of the core philosophy of the future framework---as a developer you can never know what future backends the user have access to. Moreover, by not making any assumptions about what backends are available, your code will also work automatically with any new backends developed after you wrote your code.

source: https://future.futureverse.org/reference/plan.html#:~:text=recommended%20for%20this.-,For%20package%20developers,-Please%20refrain%20from

Reply:
I have no experience in those stuff but for now I consider future and ExtendedTask as alternatives - i.e. you choose just one of them. Here, you use both. Is that correct?

Reply:
@pawelru `shiny::ExtendedTask$new()` takes a function as an argument, but this function needs to return a `promise` or `future` object. So `shiny::ExtendedTask` is based on `future`. And it is created so that the usage of asynchronus programming in shiny is proper.

<img width="368" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/5e52b707-f866-4b08-a80f-b6baba77a218">

<img width="656" alt="image" src="https://github.com/insightsengineering/teal/assets/133694481/7c4ca3c2-87f4-445e-86b8-f27703af0c89">


So `shiny::ExtendedTask` is better than `future` alone but is based on `future`.

Reply:
This question

> is this really necessary?

If you do not set

> future::plan(future::multisession, workers = 2)

then you have one thread and the app hangs to execute the `$invoke()` (invoking of shiny::ExtendedTask()`.
If you set a plan that allows to have 2 threads, then the ExtendedTask is set on another worker, where main worker handles shiny regular session.

Reply:
I know that parallel setup inside the code is not recommended. That's why I wanted to go with the proposition of @vedhav stated in here https://github.com/insightsengineering/teal/pull/1224#issuecomment-2124913037 - just to let the user specify the lockfile. No, if we want to remove the `plan::future` outside of the package, user will also need to setup `plan::future` on his/her own, which is more complex than just creation of the lockfile.

Reply:
We ended up agreeing with @gogonzo that we want to run this asynchronously with the parallel backend that uses `future::plan`

Reply:
Hey, for the parallel asynchronous background: if there was already a `future::plan` running in the background we reuse it. If there was no parallel plan, then we create a new one with 2 workers, and we close it as soon as possible.